### PR TITLE
Add logging calls to NCF

### PR DESCRIPTION
**Pull request by robieta**
This PR adds a util for mlperf logging (conditional logging and shielded import), and adds logging calls to NCF.

### CODE REVIEW

File: official/recommendation/data_async_generation.py [hunk 1]
```diff
@@ -276,6 +292,8 @@ def _construct_records(
     if num_padding:
       # In order to have a full batch, randomly include points from earlier in
       # the batch.
+
+      mlperf_helper.ncf_print(key=mlperf_helper.TAGS.INPUT_ORDER)
```
[the remark this answers is not in the document]
Comment (robieta) [reply official/recommendation/data_async_generation.py]: I put it anywhere that affects the input order.

File: official/recommendation/data_async_generation.py [hunk 2]
```diff
@@ -287,10 +305,20 @@ def _construct_records(
...
+  if is_training:
+    # The number of points is slightly larger than num_pts due to padding.
+    mlperf_helper.ncf_print(key=mlperf_helper.TAGS.INPUT_SIZE,
+                            value=int(data[0].shape[0]))
+    mlperf_helper.ncf_print(key=mlperf_helper.TAGS.INPUT_BATCH_SIZE,
+                            value=batch_size)
+  else:
+    # num_pts is logged instead of int(data[0].shape[0]), because the size
+    # of the data vector includes zero pads which are ignored.
+    mlperf_helper.ncf_print(key=mlperf_helper.TAGS.EVAL_SIZE, value=num_pts)
```
[the remark this answers is not in the document]
Comment (robieta) [reply official/recommendation/data_async_generation.py]: Training actually uses the padded points, but eval drops them.

File: official/recommendation/data_preprocessing.py [hunk 3]
```diff
@@ -147,6 +151,12 @@ def _filter_index_sort(raw_rating_path, match_mlperf):
   num_users = len(original_users)
   num_items = len(original_items)
 
+  mlperf_helper.ncf_print(key=mlperf_helper.TAGS.PREPROC_HP_NUM_EVAL,
```
[the remark this answers is not in the document]
Comment (robieta) [reply official/recommendation/data_preprocessing.py]: "PREPROC_" tags appear before the clock start.

File: official/recommendation/data_preprocessing.py [hunk 4]
```diff
@@ -450,6 +461,12 @@ def instantiate_pipeline(dataset, data_dir, batch_size, eval_batch_size,
       "Wrote flagfile for async data generation in {}."
       .format(flagfile))
 
+
+  # The async process performs the run specific random calls, therefore the
+  # clock is started before instantiating the subprocess.
+  mlperf_helper.clear_system_caches()
+  mlperf_helper.ncf_print(key=mlperf_helper.TAGS.RUN_START)
```
[the remark this answers is not in the document]
Comment (robieta) [reply official/recommendation/data_preprocessing.py]: I did a refactor which moves the flagfile write to the very end. That way both processes start their respective tasks at the same time.

File: official/utils/logs/mlperf_helper.py [hunk 5]
```diff
@@ -0,0 +1,233 @@
...
+  version_str = "{}.{}.{}".format(*parsed_line.version)
+  callsite_str = "({}:{})".format(*parsed_line.callsite)
+  value_str = ": {}".format(parsed_line.value) if parsed_line.value else ""
+  return ":::MLPv{} {} {} {} {} {}".format(
+      version_str, parsed_line.benchmark, parsed_line.timestamp, callsite_str,
+      parsed_line.tag, value_str)
+
+
+def get_mlperf_log():
+  """Shielded import of mlperf_log module."""
```
[the remark this answers is not in the document]
Comment (robieta) [reply official/utils/logs/mlperf_helper.py]: People don't follow requirements. (Also we have a non-trivial number of Conda users) No sense in breaking people for no good reason.

File: official/utils/logs/mlperf_helper.py [hunk 6]
```diff
@@ -0,0 +1,233 @@
...
+
+class Logger(object):
+  """MLPerf logger indirection class.
+
+  This logger only logs for MLPerf runs, and prevents various errors associated
+  with not having the mlperf_compliance package installed.
+  """
+  class Tags(object):
+    def __init__(self, mlperf_log):
+      self._enabled = False
```
[the remark this answers is not in the document]
Comment (robieta) [reply official/utils/logs/mlperf_helper.py]: Fixed.

File: official/utils/logs/mlperf_helper.py [hunk 7]
```diff
@@ -0,0 +1,242 @@
...
+
+SUDO = "sudo" if os.geteuid() else ""
+
+# This indirection is used in docker.
+DROP_CACHE_LOC = os.getenv("DROP_CACHE_LOC", "/proc/sys/vm/drop_caches")
+
+_NCF_PREFIX = "NCF_RAW_"
+
+# TODO(robieta): move line parsing to mlperf util
+_PREFIX = r"({})?:::MLPv([0-9]+).([0-9]+).([0-9]+)".format(_NCF_PREFIX)
```
Comment (reedwm) [inline official/utils/logs/mlperf_helper.py]: I don't think the first group needs to be a capturing group. So you can replace `({})?` with `(?:{})?`, then change other code to take into account this group isn't captured.

Comment (robieta) [reply official/utils/logs/mlperf_helper.py]: Thanks.

**Pull request merged**